File: hash/js/rosetta-squint-hash/src/internal/lanczos.ts
/**
 * Pillow-compatible Lanczos3 resize on uint8 grayscale (Uint8Array, row-major).
 *
 * Reproduces libImaging/Resample.c precompute_coeffs precisely:
 *   - center = (idx + 0.5) * scale
 *   - filter_scale = max(1.0, scale)
 *   - support = 3.0 * filter_scale
 *   - kernel = sinc(x) * sinc(x/3) for |x| < 3
 *   - xmin = trunc(center - support + 0.5), clamped to [0, src_size)
 *   - xmax = trunc(center + support + 0.5), clamped to (xmin, src_size] EXCLUSIVE
 *   - weights normalized per output pixel
 *   - PRECISION_BITS = 32 - 8 - 2 = 22 for fixed-point coefficients (NOT 32)
 *
 * IMPORTANT JS-specific note: the accumulator can exceed 2^31, so we CANNOT use
 * bitwise `>>>` (which truncates to int32). Use float arithmetic + Math.floor.
 */

const PRECISION_BITS = 32 - 8 - 2; // = 22 — matches Pillow's #define PRECISION_BITS
const SUPPORT = 3.0;
const PRECISION_SCALE = 1 << PRECISION_BITS; // 4194304, fits in int32

export function resize(
  src: Uint8Array,
  srcW: number,
  srcH: number,
  dstW: number,
  dstH: number,
): Uint8Array {
  const { offsets: offsH, lengths: lensH, weights: weightsH } = precomputeCoeffs(srcW, dstW);
  const mid = new Uint8Array(srcH * dstW);
  for (let y = 0; y < srcH; y++) {
    const rowOff = y * srcW;
    for (let xd = 0; xd < dstW; xd++) {
      let acc = 0;
      const w = weightsH[xd];
      const off = offsH[xd];
      const len = lensH[xd];
      for (let i = 0; i < len; i++) {
        acc += w[i] * src[rowOff + off + i];
      }
      mid[y * dstW + xd] = clip8(acc);
    }
  }

  const { offsets: offsV, lengths: lensV, weights: weightsV } = precomputeCoeffs(srcH, dstH);
  const result = new Uint8Array(dstH * dstW);
  for (let yd = 0; yd < dstH; yd++) {
    const w = weightsV[yd];
    const off = offsV[yd];
    const len = lensV[yd];
    for (let x = 0; x < dstW; x++) {
      let acc = 0;
      for (let i = 0; i < len; i++) {
        acc += w[i] * mid[(off + i) * dstW + x];
      }
      result[yd * dstW + x] = clip8(acc);
    }
  }
  return result;
}

function clip8(acc: number): number {
  // Math.floor matches Java's arithmetic right shift for both positive and negative acc
  // (rounds toward -infinity). Adding (1 << (PRECISION_BITS - 1)) before division
  // is the round-half-up bias used by Pillow.
  const rounded = Math.floor((acc + (1 << (PRECISION_BITS - 1))) / PRECISION_SCALE);
  if (rounded < 0) return 0;
  if (rounded > 255) return 255;
  return rounded;
}

interface CoeffTable {
  offsets: number[];
  lengths: number[];
  weights: Int32Array[]; // per dst pixel, length = lengths[i]
}

function precomputeCoeffs(srcSize: number, dstSize: number): CoeffTable {
  const scale = srcSize / dstSize;
  const filterScale = Math.max(1.0, scale);
  const support = SUPPORT * filterScale;

  const offsets: number[] = new Array(dstSize);
  const lengths: number[] = new Array(dstSize);
  const weights: Int32Array[] = new Array(dstSize);

  for (let xd = 0; xd < dstSize; xd++) {
    const center = (xd + 0.5) * scale;
    // Math.trunc matches Pillow's `(int)(x + 0.5)` for non-negative x (truncates toward zero).
    let xmin = Math.trunc(center - support + 0.5);
    if (xmin < 0) xmin = 0;
    let xmax = Math.trunc(center + support + 0.5);
    if (xmax > srcSize) xmax = srcSize;
    const n = Math.max(0, xmax - xmin);

    const tmp = new Float64Array(n);
    let wsum = 0;
    for (let i = 0; i < n; i++) {
      const dx = ((xmin + i) + 0.5 - center) / filterScale;
      const w = lanczosKernel(dx);
      tmp[i] = w;
      wsum += w;
    }
    if (wsum !== 0) {
      for (let i = 0; i < n; i++) tmp[i] /= wsum;
    }
    const q = new Int32Array(n);
    for (let i = 0; i < n; i++) {
      q[i] = Math.round(tmp[i] * PRECISION_SCALE);
    }
    offsets[xd] = xmin;
    lengths[xd] = n;
    weights[xd] = q;
  }
  return { offsets, lengths, weights };
}

function lanczosKernel(x: number): number {
  if (x === 0) return 1;
  const ax = Math.abs(x);
  if (ax >= SUPPORT) return 0;
  const px = Math.PI * x;
  return (Math.sin(px) / px) * (Math.sin(px / SUPPORT) / (px / SUPPORT));
}

File: hash/rust/rosetta-squint-hash/src/internal/lanczos.rs
//! Pillow-compatible Lanczos3 resize on uint8 grayscale (Vec<Vec<u8>> in [0,255]).
//!
//! Reproduces libImaging/Resample.c precompute_coeffs precisely:
//!   - center = (idx + 0.5) * scale
//!   - filter_scale = max(1.0, scale)
//!   - support = 3.0 * filter_scale
//!   - kernel = sinc(x) * sinc(x/3) for |x| < 3
//!   - xmin = (center - support + 0.5) as i32, clamped to [0, src_size)
//!   - xmax = (center + support + 0.5) as i32, clamped to (xmin, src_size] EXCLUSIVE upper bound
//!   - weights normalized per output pixel
//!   - PRECISION_BITS = 32 - 8 - 2 = 22 for fixed-point coefficients (NOT 32; 1<<32 overflows i32)
//!   - acc accumulated in i64

const PRECISION_BITS: u32 = 32 - 8 - 2; // = 22 — matches Pillow's #define PRECISION_BITS
const SUPPORT: f64 = 3.0;

/// Resize src[H][W] uint8 to a new [dst_h][dst_w] uint8 buffer via Lanczos3.
pub fn resize(src: &[Vec<u8>], dst_w: usize, dst_h: usize) -> Vec<Vec<u8>> {
    let src_h = src.len();
    let src_w = src[0].len();

    let (offs_h, lens_h, weights_h) = precompute_coeffs(src_w, dst_w);
    let mut mid: Vec<Vec<u8>> = Vec::with_capacity(src_h);
    for y in 0..src_h {
        let row = &src[y];
        let mut out = vec![0u8; dst_w];
        for xd in 0..dst_w {
            let mut acc: i64 = 0;
            let w = &weights_h[xd];
            let off = offs_h[xd];
            for i in 0..lens_h[xd] {
                acc += i64::from(w[i]) * i64::from(row[off + i]);
            }
            out[xd] = clip8(acc);
        }
        mid.push(out);
    }

    let (offs_v, lens_v, weights_v) = precompute_coeffs(src_h, dst_h);
    let mut result: Vec<Vec<u8>> = Vec::with_capacity(dst_h);
    for yd in 0..dst_h {
        let w = &weights_v[yd];
        let off = offs_v[yd];
        let mut out = vec![0u8; dst_w];
        for x in 0..dst_w {
            let mut acc: i64 = 0;
            for i in 0..lens_v[yd] {
                acc += i64::from(w[i]) * i64::from(mid[off + i][x]);
            }
            out[x] = clip8(acc);
        }
        result.push(out);
    }
    result
}

fn clip8(acc: i64) -> u8 {
    let rounded = (acc + (1_i64 << (PRECISION_BITS - 1))) >> PRECISION_BITS;
    rounded.clamp(0, 255) as u8
}

fn precompute_coeffs(src_size: usize, dst_size: usize) -> (Vec<usize>, Vec<usize>, Vec<Vec<i32>>) {
    let scale = src_size as f64 / dst_size as f64;
    let filter_scale = scale.max(1.0);
    let support = SUPPORT * filter_scale;

    let mut offsets = vec![0usize; dst_size];
    let mut lengths = vec![0usize; dst_size];
    let mut weights: Vec<Vec<i32>> = Vec::with_capacity(dst_size);

    for xd in 0..dst_size {
        let center = (xd as f64 + 0.5) * scale;
        let xmin_f = center - support + 0.5;
        let xmax_f = center + support + 0.5;
        let mut xmin = xmin_f as i32;
        if xmin < 0 {
            xmin = 0;
        }
        let mut xmax = xmax_f as i32;
        if xmax > src_size as i32 {
            xmax = src_size as i32;
        }
        let n = (xmax - xmin).max(0) as usize;
        let xmin = xmin as usize;

        let mut tmp = vec![0.0_f64; n];
        let mut wsum = 0.0;
        for i in 0..n {
            let dx = ((xmin + i) as f64 + 0.5 - center) / filter_scale;
            let w = lanczos_kernel(dx);
            tmp[i] = w;
            wsum += w;
        }
        if wsum != 0.0 {
            for w in tmp.iter_mut() {
                *w /= wsum;
            }
        }
        let mut q = vec![0i32; n];
        for i in 0..n {
            q[i] = (tmp[i] * ((1u32 << PRECISION_BITS) as f64)).round() as i32;
        }
        offsets[xd] = xmin;
        lengths[xd] = n;
        weights.push(q);
    }
    (offsets, lengths, weights)
}

fn lanczos_kernel(x: f64) -> f64 {
    if x == 0.0 {
        return 1.0;
    }
    let ax = x.abs();
    if ax >= SUPPORT {
        return 0.0;
    }
    let px = std::f64::consts::PI * x;
    (px.sin() / px) * ((px / SUPPORT).sin() / (px / SUPPORT))
}
